rs/tests/execution/general_execution_tests/nns_shielding.rs

```diff
@@ -141,6 +142,103 @@ pub fn mint_cycles_not_supported_on_application_subnet(env: TestEnv) {
+            .await
+            .unwrap()
+            .canister_id();
+
+        let before_balance = get_balance(&not_cmc, &nns_agent).await;
+        assert_balance_equals(
+            *INITIAL_CYCLES,
+            Cycles::from(before_balance),
+            BALANCE_EPSILON,
+        );
+
+        let res = nns_agent
+            .update(&not_cmc, "update")
+            .with_arg(
+                wasm()
+                    .mint_cycles128(Cycles::from((1u128 << 64) + 2u128))
+                    .reply_data_append()
+                    .reply()
+                    .build(),
+            )
+            .call_and_wait()
+            .await
+            .expect_err("should not succeed");
+
+        assert_eq!(
+            res,
+            AgentError::CertifiedReject(
+                RejectResponse {
+                    reject_code: RejectCode::CanisterError,
+                    reject_message: format!(
+                        "Error from Canister {}: Canister violated contract: ic0.mint_cycles cannot be executed on non Cycles Minting Canister: {} != {}.\nThis is likely an error with the compiler/CDK toolchain being used to build the canister. Please report the error to IC devs on the forum: https://forum.dfinity.org and include which language/CDK was used to create the canister.",
+                        not_cmc, not_cmc,
+                        CYCLES_MINTING_CANISTER_ID),
+                    error_code: None})
+        );
+
+        let after_balance = get_balance(&not_cmc, &nns_agent).await;
+        assert!(
+            after_balance == before_balance,
+            "expected {} == {}",
+            after_balance,
+            before_balance
+        );
+    });
+}
+
+pub fn mint_cycles128_not_supported_on_application_subnet(env: TestEnv) {
+    let initial_cycles = CANISTER_FREEZE_BALANCE_RESERVE + Cycles::new(5_000_000_000_000);
+    let app_node = env.get_first_healthy_application_node_snapshot();
+    let agent = app_node.build_default_agent();
+    block_on(async move {
+        let canister_id = UniversalCanister::new_with_cycles(
+            &agent,
+            app_node.effective_canister_id(),
+            initial_cycles * 3u64,
+        )
+        .await
+        .unwrap()
+        .canister_id();
+
+        let before_balance = get_balance(&canister_id, &agent).await;
+        assert!(
+            Cycles::from(before_balance) > initial_cycles * 2u64,
+            "expected {} > {}",
+            before_balance,
+            initial_cycles * 2u64
+        );
+        assert!(
+            Cycles::from(before_balance) <= initial_cycles * 3u64,
+            "expected {} <= {}",
+            before_balance,
+            initial_cycles * 3u64
+        );
+
+        // The test function on the wasm module will call the mint_cycles system
+        // call.
+        let res = agent.update(&canister_id, "test").call_and_wait().await;
```

Comment:
Why not use the universal canister library as in the above test for NNS?

Reply:
because of the test template (mint_cycles tests). Is there any value in making them uniform?

Reply:
Once there are uniform, one can factor out the common port to avoid code duplication.

Reply:
Which test to you mean? The two old tests using the MINT_CYCLES.wat?

Reply:
> MINT_CYCLES.wat

ideally, this is not needed at all (the universal canister is supposed to be used instead)

Reply:
But they do use an universal canister. 

>as in the above test for NNS?

Which one?

Reply:
I'm not saying the tests should be disabled: I'm suggesting to avoid code duplication by making them as uniform as possible.

Reply:
I don't understand what you'd like me to change, or what your first comment refers to. 

The two new *128* tests are here to stay, and they use the universal canister. If there is something else about them that needs attention, please let me know. The other mint_cycles tests in this file are going to be deprecated in a short while, so making them similar to the new tests is not worth it.

Reply:
The test `mint_cycles128_supported_only_on_cycles_minting_canister` invokes the universal canister as:
```
        let res = nns_agent
            .update(&not_cmc, "update")
            .with_arg(
                wasm()
                    .mint_cycles128(Cycles::from(2u128))
                    .reply_data_append()
                    .reply()
                    .build(),
            )
            .call_and_wait()
            .await
            .expect_err("should not succeed");
```
while the test `mint_cycles128_not_supported_on_application_subnet` invokes the universal canister as:
```
let res = agent.update(&canister_id, "test").call_and_wait().await;
```
Given that the universal canister doesn't export a method called "test", I suppose the line `assert_reject(res, RejectCode::CanisterError);` succeeds because the calls fails due to no method called "test" being exported by the universal canister.

Does it make more sense now? I'm also happy to set up a quick call to discuss this interactively.

Reply:
My suggestion is to use the former pattern in both tests and then also factor out the common part into a helper function, e.g., setting up the universal canister, making the update call, and returning the update call's result; then the tests could have the form:
```
fn setup_ucan_and_call_mint_cycles(env: TestEnv) {
  ...
  nns_agent
            .update(&not_cmc, "update")
            .with_arg(
                wasm()
                    .mint_cycles128(Cycles::from(2u128))
                    .reply_data_append()
                    .reply()
                    .build(),
            )
            .call_and_wait()
            .await
            .expect_err("should not succeed")
}

pub fn test_1(env: TestEnv) {
  let (res, balance_diff) = setup_ucan_and_call_mint_cycles(env);
  assert_eq!(
            res,
            AgentError::CertifiedReject(...)
  );
  assert_eq!(balance_diff, 0);
}

pub fn test_2(env: TestEnv) {
  let (res, balance_diff) = setup_ucan_and_call_mint_cycles(env);
  assert_reject(res, RejectCode::CanisterError);
  assert!(balance_diff < EPSILON);
}
```